Stig shadow requests

TheStig294 opened the pull request:
- Changed messages displayed to shadow if the join team buff is active
- Join team buff is now the default buff
- Added convars for the shadow and target to be soul-linked,
the shadow to copy the role of their target rather than become a basic team member role,
and whether credit is given to killing a soul-linked player for purposes of things like a jester winning
- Added help description if soul link is turned on for the shadow

Malivil commented on gamemodes/terrortown/gamemode/roles/shadow/shadow.lua:
```diff
@@ -90,7 +94,12 @@ local function ClearBuffTimer(shadow, target, sendMessage)
         if sendMessage then
-            local message = "You got too far from your target and stopped buffing them!"
+            local message = "You got too far from your target and "
+            if target_buff:GetInt() == 4 then
```
Use the constant (SHADOW_BUFF_TEAM_JOIN)

Malivil commented on gamemodes/terrortown/gamemode/roles/shadow/shadow.lua:
```diff
@@ -117,7 +126,12 @@ local function CreateBuffTimer(shadow, target)
     local buffDelay = target_buff_delay:GetInt()
-    local message = "Stay with your target for " .. buffDelay .. " seconds to give them a buff!"
+    local message = "Stay with your target for " .. buffDelay .. " seconds to "
+    if target_buff:GetInt() == 4 then
```
Use the constant here too

Malivil commented on gamemodes/terrortown/gamemode/roles/shadow/shadow.lua:
```diff
@@ -196,6 +211,58 @@ hook.Add("ScalePlayerDamage", "Shadow_Buff_ScalePlayerDamage", function(ply, hit
+
+    local killDmg = DamageInfo()
+    killDmg:SetDamage(ply:Health() + 1)
+    killDmg:SetDamageType(dmg:GetDamageType())
```
This won't work if the player has something that scales down damage

Malivil commented on gamemodes/terrortown/gamemode/roles/shadow/shadow.lua:
```diff
@@ -196,6 +211,58 @@ hook.Add("ScalePlayerDamage", "Shadow_Buff_ScalePlayerDamage", function(ply, hit
+        end
+    else
+        -- Find the shadows that "belong" to this player, and kill them
+        for _, p in ipairs(player.GetAll()) do
```
Use the cached `GetAllPlayers`

Malivil reviewed (changes requested):
Also missing changes to release notes, convars, and version number

TheStig294 commented:
I couldn't get the soul link death credit convar to work so I just removed it. The kill command should work every time.
Also, I noticed the shadow wasn't dying if they killed their target so I changed a line in the PlayerDeath hook.

Malivil commented on gamemodes/terrortown/gamemode/roles/shadow/shadow.lua:
```diff
@@ -305,7 +349,6 @@ end)
 hook.Add("PlayerDeath", "Shadow_KillCheck_PlayerDeath", function(victim, infl, attacker)
     local valid_kill = IsPlayer(attacker) and attacker ~= victim and GetRoundState() == ROUND_ACTIVE
     if not valid_kill then return end
-    if attacker:IsShadow() then return end
```
Revert this.  This makes it so the shadow always dies when their target is killed not just when the shadow kills them. 

The bug you mentioned only happens if the attacker isn't set correctly, like when using the shark trap

NoxxFlame commented:
LGTM
I'll merge this and then push a beta update
